Add Icon and Media Support to List  - Resolves #906 and #2272 (#2316)
MarkBind/markbind: docs/userGuide/syntax/lists.md

</variable>
</include>

****Customizing the Unordered list appearance:****

**To customize Unordered list, add the configuration `{icon="icon-name"}` after a specific list item**

<include src="codeAndOutput.md" boilerplate >
<variable name="highlightStyle">markdown</variable>
<variable name="code">
* Item 1 { icon="glyphicon-education" }
* Item 2
  * Item 2.1 { icon="fas-file-code" }
  * Item 2.2
* Item 3 { icon="fas-code-branch" }
  * Item 3.1 
* Item 4 { icon="octicon-git-pull-request" }
  * Item 4.1 { icon="mif-perm-media" }
* Item 5 { icon="glyphicon-education" }
  * Item 5.1 { icon="notebook_with_decorative_cover" }
  </variable>
</include>

1. If an item has a specified icon, that icon will be used for it and for subsequent items at that level.
2. If any item at a level is customized, the first item at that level must also be customized. If not, the list will be invalidated.
3. You can use any of the [icons](../formattingContents.html#icons) supported by MarkBind.

> 4. When using icons, omit the `::` around the icon name (e.g., use `{icon="fas-home"}` instead of `{icon=":fas-home:"}`).

Reviewer: `invalidated` is also a fairly broad technical term (e.g. what's being invalidated? will the entire list disappear?)

Try to be more specific and use layman language, e.g. `If not, the customizations will be ignored`.

> 4. When using icons, omit the `::` around the icon name (e.g., use `{icon="fas-home"}` instead of `{icon=":fas-home:"}`).

Reviewer: Sharing a read, https://www.stylemanual.gov.au/writing-and-designing-content/clear-language-and-writing-style/plain-language-and-word-choice

This article is more ui/ux focused but I find many of its points to be helpful in technical writing and day to day communication as well.

**You can adjust the icon's size by using the `i-size` attribute.**

<include src="codeAndOutput.md" boilerplate >
<variable name="highlightStyle">markdown</variable>
<variable name="code">

* Item 1 {icon="fas-file-code" i-size="35px"}
* Item 2 {icon="fas-file-code" i-size="4rem"}
* Item 3 {icon="fas-file-code" i-size="5em"}

</variable>
</include>

You can utilize any [CSS size unit](https://www.w3schools.com/cssref/css_units.php).

**You can use images as icon.**

<include src="codeAndOutput.md" boilerplate >
<variable name="highlightStyle">markdown</variable>
<variable name="code">

* Item 1 {icon="/images/deer.jpg" i-width="30px"}
* Item 2 {icon="/images/deer.jpg" i-width="60px" i-height="34px"}
* Item 3 {icon="/images/deer.jpg" i-width="90px" i-height="51px"}
* Item 4 {icon="/images/deer.jpg" i-width="120px" i-height="68px"}

> * Item 5 {icon="/images/deer.jpg" i-width="150px" i-height="84px"}

Reviewer: why repeat `icon` in this example?

</variable>
</include>

If either the `i-width` or the `i-height` of an image is not specified, the unspecified dimension will adjust to maintain the image's original aspect ratio (e.g. Original Image: 800x600 (aspect ratio 4:3), Set `i-width`: 500px, Outcome: 500x375).

**The icon's appearance can be further customized by adding a `i-class` attribute.**

<div id="main-example-markbind">

<include src="codeAndOutput.md" boilerplate >
<variable name="highlightStyle">markdown</variable>
<variable name="code">
* Item 1 <br>
  Item 1 line 2{icon="/images/deer.jpg" i-width="60px" height="17px" i-class="rounded"}
* Item 2
  Item 2 Continue

>   * Item 2.1 {icon="fas-question-circle" i-class="badge rounded-pill mb-1 bg-success text-white"}

Reviewer: looks a little tight

```suggestion
  * Item 2.1 {icon="fas-question-circle" i-class="badge rounded-pill my-1 bg-success text-white"}
```

(and the other `mb-1`'s too)

<include src="codeAndOutput.md" boilerplate >
<variable name="highlightStyle">markdown</variable>
<variable name="code">

* #### Heading 1: Overview {icon="/images/overview-d.png" i-width="65px" i-class="rounded" }
  
   Content 1: This section provides a summary of the document or topic. 
   It sets the context and purpose of the content to follow.

   * ##### Heading 1.1: Highlights { icon="/images/highlights-d.png" i-width="65px" i-class="rounded" }
  
      Content 1.1: Here we summarize the key points, findings, or features of the main topic. 
      This could include important data, outcomes, or insights.

* #### Heading 2: Detailed Description { icon="/images/detailed-d.png" i-width="65px" i-class="rounded" }
    
   Content 2: This section delves deeper into the topic, offering comprehensive information and detailed explanations.
   It might also include evidence, examples, or justifications.

* #### Heading 3: Conclusion { icon="/images/conclusion-d.png" i-width="65px" i-class="rounded" }
  
    Content 3: The conclusion draws together the main threads of the topic, 
    summarizing the central points and their implications.

   * ##### Heading 3.1: Final Thoughts
  

>      Content 3.1: This is a place for final reflections, recommendations, or a look towards future developments or trends.

Reviewer: similarly, this example could be more concise for the reader. (e.g. heading 1.1, 3, 3.1 seems unnecessary)

Reviewer: this was added back in your revert commit

</variable>
</include>

1. Icon specifications should be attached only to the first element of a list item (for the example above, the icon specification should be attached to the heading, not the block of text below the heading).

> 2. A blank line must separate a heading from the content that follows it.

Reviewer: Is `2.` required? Was able to get it working without:

![image](https://github.com/MarkBind/markbind/assets/3306138/9fd7f295-6823-4f9a-8ce9-6c8e2d25be0b)

Reviewer: might have deleted the wrong one

Author: Oops fixed.
2. is not required.

- [x] Item 4
- ( ) Item 5
</div>

> 

Reviewer 2: It would be nice to use images that match with each other e.g., similar style, color palette, etc.

> 

Author: All images are AI-generated, so there's no need to worry about copyright issues.

****Customizing the list appearance:****

**To customize list icons, add the configuration `{icon="icon-name"}` either after a specific list item or at the end of the entire list.**

<include src="codeAndOutput.md" boilerplate >
<variable name="highlightStyle">markdown</variable>
<variable name="code">
* Item 1 {icon="glyphicon-education"}
* Item 2
  * Item 2.1 {icon="fas-file-code"}
  * Item 2.2
* Item 3 {icon="fas-code-branch"}
  * Item 3.1 
* Item 4 {icon="octicon-git-pull-request"}
  * Item 4.1 {icon="mif-perm-media"}
* Item 5 {icon="glyphicon-education"}
  * Item 5.1 {icon="notebook_with_decorative_cover"}
  </variable>
</include>

> <box type="definition" seamless>
> 
> 1. You can use any of the icons supported by MarkBind (e.g., Font Awesome, Octicons, Glyphicons, Images, etc.) as the icon of a list item.
> 1. When using emojis, omit the `::` around the icon name (e.g., use `{icon="emoji"}` instead of `{icon=":emoji:"}`).
> 1. If an item has a specified icon, that icon will be used for it and for subsequent items at that level. 
> 1. If any item at a level is customized, the first item at that level must also be customized. If not, the list will be invalidated.
> </box>

Reviewer 2: ```suggestion
1. You can use any of the [icons (e.g., Font Awesome, Octicons, Glyphicons, etc.)](#icons) or [emojis](#emoji) supported by MarkBind as the icon of a list item, but omit the `:` at the start and at the end .e.g, `fas-home` instead of `:fas-home:`.
1. If an item has a specified icon, that icon will be used for it and for subsequent items at that level e.g., `Item 2` in the example list above.
1. If any item at a level is customized, the first item at that level must also be customized. If not, the list will be considered invalid.
```

1. These are not definitions. So, `info` is a better match. 
2. Furthermore, best not to put inside a box at all so that we can use boxes for 'special occasions' only. Also, to be consistent with the rest of the UG.
3. Images are not an 'icon supported by MarkBind'

**The icon's appearance can be further customized by adding a `class` attribute.**

<div id="main-example-markbind">

<include src="codeAndOutput.md" boilerplate >
<variable name="highlightStyle">markdown</variable>
<variable name="code">
* Item 1 <br>
  Item 1 line 2  {icon="fas-file-code" class="text-warning"}
* Item 2 
  Item 2 line 1 continue 
  * Item 2.1 {icon="fas-code-branch" class="text-success"}
  * Item 2.2 {class="text-white"}
* Item 3

>   * Item 3 {icon="/images/deer.jpg" width="30px" height="17px" class="rounded text-white"}

Reviewer 2: `text-white` doesn't make sense for an image? The effect of `rounded` is not visible to the naked eye either.
Also, we have not introduced `width`/`height` attributes yet.

</variable>
</include>

<box type="definition" seamless>

> Each item level attribute supersedes the default one. For instance, the class attribute of item 2.2 overrides the default attribute of item 2.1.

Reviewer 2: Let's not use the term `item level`, as we don't have a `list-level`. Instead, we can call them `icon specification`.
```suggestion
Once a `class` attribute is specified for an icon, it will continue to apply for icons at the same level, until overridden by another `class` attribute.
```

Reviewer 2: The example can be tweaked to illustrate the waterfall behavior of the `class` attribute.

</box>

</div>

> **You can adjust the icon's size by using the `size` attribute.**

Reviewer 2: Do we even have a `size` attribute?

<include src="codeAndOutput.md" boilerplate >
<variable name="highlightStyle">markdown</variable>
<variable name="code">

* Item 1 {icon="/images/deer.jpg" width="30px" height="17px"}
* Item 2 {icon="/images/deer.jpg" width="60px" height="34px"}
* Item 3 {icon="/images/deer.jpg" width="90px" height="51px"}
* Item 4 {icon="/images/deer.jpg" width="120px" height="68px"}
* Item 5 {icon="/images/deer.jpg" width="150px" height="84px"}

</variable>
</include>

<box type="definition" seamless>

> 1. All CSS units are acceptable for size. 

Reviewer 2: Give examples. Target users are not necessarily CSS programmers. Alternatively, point to another resource that provides a newbie-friendly explanation of CSS size syntax.

> 2. If height is omitted, width will be used instead, and vice versa. If both are omitted, the size defaults to the smallest icon size.

Reviewer 2: This is not clear. In a normal image, if only one dimension is specified, the whole image will be scaled to preserve original proportions. If none are specified, the original size is used. Are we deviating from both both of those behaviors?

</box>

**You can apply Markdown's heading and paragraph syntax within the list.**

<include src="codeAndOutput.md" boilerplate >
<variable name="highlightStyle">markdown</variable>
<variable name="code">

* ## Heading 1: MarkBind Overview {icon="/images/list.png" width="65px" class="rounded"}

>   

Reviewer 2: Consider using a lower-level heading so that they don't appear in the pageNav .

> * ## Heading 1: MarkBind Overview {icon="/images/list.png" width="65px" class="rounded"}
>   
>   Content 1: MarkBind is a tool for generating dynamic, text-heavy websites from Markdown. It supports various syntax schemes for enhanced content dynamism.
> 
> * ## Heading 2 Features {icon="/images/toolbox.png" width="65px" class="rounded"}
>   
>   Content 2: MarkBind offers built-in features like icons, emoji, and search functionality. It also supports content reuse and multiple organization methods.
> 
> * ## Heading 3 Setup and Integration {icon="/images/setup.png" width="65px" class="rounded"}
>   
>   Content 3: MarkBind is easy to install, modify, and deploy. It integrates seamlessly with software project workflows and offers a live preview feature.

Reviewer 2: ```suggestion
* ## Heading 1: MarkBind Overview {icon="/images/list.png" width="65px" class="rounded"}
  
  Content 1: MarkBind is a tool for generating dynamic, text-heavy websites from Markdown.
  It supports various syntax schemes for enhanced content dynamism.

* ## Heading 2 Features {icon="/images/toolbox.png" width="65px" class="rounded"}
  
  Content 2: MarkBind offers built-in features like icons, emoji, and search functionality.
  It also supports content reuse and multiple organization methods.

* ## Heading 3 Setup and Integration {icon="/images/setup.png" width="65px" class="rounded"}
  
  Content 3: MarkBind is easy to install, modify, and deploy.
  It integrates seamlessly with software project workflows and offers a live preview feature.
```

Added line breaks to avoid horizontal scrolling in the rendered page.

Reviewer 2: Also, don't go 'meta' with examples. Doing so can make them harder to understand for the unsuspecting reader. For example, the content of this example should not be related to MarkBind at all, so that it is clear they are sample/dummy content and not actual UG content.

</variable>
</include>

<box type="definition">

> 1. Item-level specifications should be attached only to the list heading or list item, not the list content.

Reviewer 2: ```suggestion
1. Icon specifications should be attached only to the first element of a list item (for the example above, the icon specification should be attached to the heading, not the block of text below the heading).
```
Because it's not clear what is meant by `list heading` and `list content` ...

> 2. Ensuring a blank line precedes the content is crucial.

Reviewer 2: This `Ensuring ... is crucial` way of phrasing seems not direct enough to me.
```suggestion
2. A blank line must separate a heading from the content that follows it. 
```

- [x] Item 4
- ( ) Item 5
</div>

****Customizing the list appearance:****

> <box type="tip" seamless>

Reviewer 2: This is not a tip, but regular content.

> The list icons can be customized by adding an {icon="..."} at the end of a list item or a list itself.

Reviewer 2: ```suggestion
The list icons can be customized by adding an `{icon="..."}` at the end of a list item or a list itself.
```

</box>

> 

Reviewer 2: Try to use examples that looks like realistic usage. We don't want the reader to look at an example and think "Why would I want to do _that_?"

<div id="main-example-markbind">

<include src="codeAndOutput.md" boilerplate >
<variable name="highlightStyle">markdown</variable>
<variable name="code">
* item 1 {icon="x"}
* item 2 {icon="heavy_check_mark"}
* item 3
* item 4 {icon="/images/deer.jpg"}
  
{icon="fas fa-check"}
</variable>
</include>

</div>

<box type="warning">

1. Item-level specifications should only be attached to the list heading or list item, not the list content.

> 2. A blank line preceding the content is crucial.

Author: This description is correct, but it's placed incorrectly. I'll make adjustments accordingly.

> 3. It's important to insert a blank line before list-level specifications.

Reviewer 2: Minor: Recommended to use `1. ` for all items in a numbered list so that it is easy to insert items later.
e.g.,
```
1. aaa
1. bbb
1. ccc
```

4. It is possible to any of the icons supported by MarkBind (e.g., Font Awesome, Octicons, Glyphicons, Images and etc.) as the icon of a list item.

> 5. Images will maintain a 1 rem bottom margin.

Reviewer 2: We should add appropriate padding to all elements, but that is an internal detail.

Author: Padding is only added to images to avoid disrupting the original list component's behavior. The original list component uses bullet points, which naturally create spacing. Conversely, image icons can fill an entire square's height and width, potentially leading to display issues.

<img width="182" alt="image" src="https://github.com/MarkBind/markbind/assets/30100720/cebe86ba-3d19-4049-85c4-27fad061e18f">

Furthermore, if the icon size exceeds the line height, it will yield the same issues as mentioned above.

I will refine the design to make it look nicer

Reviewer 2: @SPWwj Yes, we need to tweak the padding to make it look nice when using icons or images as bullet points. Nevertheless, it is not something we need to mention in the UG.

</box>

<box type="tip" seamless>

> It is possible to specify icons for nested levels using the icons attribute.

Reviewer 2: ```suggestion
It is possible to specify icons for nested levels using the `icons` attribute.
```
Is this even correct?

</box>
<div id="main-example-markbind">

<include src="codeAndOutput.md" boilerplate >
<variable name="highlightStyle">markdown</variable>
<variable name="code">
* item 1
* item 2 
  * item 2.1
  * item 2.2

>   {icon="x"}

Author: I forgot to mention this, but a blank line can turn the list into a loose list due to issue #2313. However, this approach allows us to bypass loose list generation when crafting nested lists.

>   {icon="x"}

Reviewer 2: So, no blank line needed for list-level specifications for nested lists, but needed for the main list (as per point 3)? It would be better if users don't need to follow such intricate rules.

>   {icon="x"}

Author: Yes, I recommend using standardized syntax that requires a blank line for all list level configurations. This approach is stable and retains all attributes due to the internal design of markdown-it-attrs, which binds attributes to the nearest block-level element. 

However, due to a design flaw in markdown-it-attrs, an inner list with list level configuration can cause the outer list to become a loose list, as indicated by issue #2313.

For now, I'm using inner list level configurations without blank lines for the demo. While it's not perfect, it works for layer 2 and 3 nested lists. Please be aware this approach may present issues for more complex nested lists due to its lack of consistency.

>   {icon="x"}

Reviewer 2: I see. Makes sense from a dev POV, but seems like a random complication from a user POV.  
Is it possible to remove the blank line requirement from _all_ levels, thus making the requirement simpler to users. In addition, we can say that this feature works only up to 3 levels of nesting.

>   {icon="x"}

Author: Not possible for most outer level:
<img width="358" alt="image" src="https://github.com/MarkBind/markbind/assets/30100720/24b45276-ff36-44e0-8a05-97e43285bbbb">
<img width="437" alt="image" src="https://github.com/MarkBind/markbind/assets/30100720/299d8868-2f3b-40bb-84a5-0fe0a63d7061">

To address this issue, I can standardize the icon list by enforcing a mandatory blank line. This can be accomplished either by resolving issue #2313 or by eliminating the padding from the 'p' tag for the outermost list level. For the latter approach, the concept of loose or tight lists would no longer apply to icon lists.

>   {icon="x"}

Reviewer 2: In that case, let's leave it as it is for now, awaiting others' comments, especially from @ang-zeyu

>   {icon="x"}

Reviewer 2: > But otherwise another alternative we can consider is:
> 
> ```
> * item 1 {icon="this will be the outer list icon for the first 3 items"}
> * item 2 a default icon
>   * item 2.1 {icon="this will be all the inner list items' icon"}
>   * item 2.2
> * item 3
> * item 4 {icon="this will be the outer list icon for the last 2 items"}
> * item 5
> ```
> 
> Wdyt?

So, the idea is to keep repeating the 'last known' icon (and settings) until it is changed, and then continue with the changed icon/settings. I like it -- as it is simpler for the user.

>   {icon="x"}

Author: This is already been implemented 😂

>   {icon="x"}

Reviewer 2: > has already already been implemented 😂

@SPWwj When? This is different from the original "item-level overrides list-level" syntax.  In this approach, there is no list-level syntax.

>   {icon="x"}

Author: This logic is exactly doing the thing you mentioned above 😂
<img width="316" alt="image" src="https://github.com/MarkBind/markbind/assets/30100720/50d49a90-8902-4099-b5ba-759a0bfe6ac6">

>   {icon="x"}

Reviewer 2: In that case, there should not be any mention of `list-level` in the UG. Example (item 3 below)

```
1. Item-level specifications should be attached only to the list heading or list item, not the list content.
2. Ensuring a blank line precedes the content is crucial.
3. Inserting a blank line before list-level specifications is important.
4. Use any of the icons supported by MarkBind (e.g., Font Awesome, Octicons, Glyphicons, Images, etc.) as the icon of a list item.
5. When using emojis, omit the `::` around the icon name (e.g., use `{icon="emoji"}` instead of `{icon=":emoji:"}`).
```

>   {icon="x"}

Author: The list-level is quite useful when users have overriding needs.
<img width="523" alt="image" src="https://github.com/MarkBind/markbind/assets/30100720/70a0d1af-4eff-497e-83f1-2b0962bf50c9">

The following syntax assumes that the entire list uses the same icon, thereby reducing customizability.

```
* item 1 {icon="this will be the outer list icon for the first 3 items"}
* item 2 a default icon
  * item 2.1 {icon="this will be all the inner list items' icon"}
  * item 2.2
* item 3
* item 4 {icon="this will be the outer list icon for the last 2 items"}
* item 5
```

>   {icon="x"}

Reviewer 2: > The following syntax assumes that the entire list uses the same icon, thereby reducing customizability.

I don't think that is the case. Instead of specifying the list-level settings at the end of the list, it takes the setting of the first item as the list-level setting. It can be changed in a subsequent item. Basically, it keeps repeating the 'last used' setting until it is changed, and then keep using the changed setting.  

But I agree that the new syntax is slightly verbose when all items except a few uses a different icon.

Original:
```
* item 1
* item 2
* item 3 {icon="x"}
* item 4

{icon="tick" size="s"}
```

New syntax for the same outcome (note the slight extra verbosity at item 4):
```
* item 1 {icon="tick" size="s"}
* item 2
* item 3 {icon="x"}
* item 4 {icon="tick"}
```

On the positive side, the new syntax is simpler and more intuitive (from a non-programmer POV) e.g., no blank lines needed, and the settings can be understood by reading from the top (without needed to jump to the end to see the default). 

But I think for typical use cases the benefit of the simpler syntax outweighs the cost of extra verbosity.

>   {icon="x"}

Author: Here's the enhanced version that preserves the overriding feature of the original syntax.

```
* item 1 {icon="tick" size="s"} //Consider if you want to override this instead.
* item 2
* item 3 {icon="x"}
* item 4 {icon="tick" default}
```

>   {icon="x"}

Reviewer 2: What's the `default` for? i.e., what's the difference between item 3 and 4?
My thinking was, as the size was never changed in items 2-4, they will all be `size="s"`. But if any item specified a different size, that size will be applied for that item onwards, until a different size is specified in a subsequent item.

>   {icon="x"}

Reviewer: > Not quite but I think this could work too. 👍

Wdyt @damithc @SPWwj 

> The following syntax assumes that the entire list uses the same icon, thereby reducing customizability.

> What's the default for?

I have the same doubt as @damithc, your current implementation should be plenty fine even without the `default` attribute.

Just make sure to remove:
- `<ul>` related implementations in code
- and list/list-item references in docs

>   {icon="x"}

Reviewer: > ~~Not quite but I think this could work too. 👍~~

On the other hand, if we have the following it becomes more verbose, but there are probably cases the other way round too...

```
* item 1 {icon="this is the exception"}
* item 2 {icon="this is the default"}
* item 3 {icon="this is the default"}
* item 4 {icon="this is the default"}
* item 5 {icon="this is the default"}
```

vs

```
* item 1 {icon="this is the exception"}
* item 2 {icon="this is the default for every item below"}
* item 3
* item 4
* item 5
```

>   {icon="x"}

Author: `Default` enhances UX. Note, lists are data, prone to change. When item 2 alters, items 2 and 3 must be updated.
```
* item 1 {icon="this is the exception"} // icon is present
* item 2 {icon="this is the default"}     // icon is present but to change to exception
* item 3 {icon="this is the default"}     // no icon need to set to default, otherwise below will be override by exception what is user forget to set?
* item 4 {icon="this is the default"}     // no icon
* item 5 {icon="this is the default"}    // no icon
```
With "default", there's no specific placement for the icon. Thus, when item 2 changes, only item 2 needs to be updated.
```
* item 1 {icon="this is the exception"} // icon is present
* item 2 {icon="this is the default"}     // no icon, but want to change to exception
* item 3 {icon="this is the default"}     // no icon
* item 4 {icon="this is the default"}     // no icon
* item 5 {icon="this is the default"}    // a default icon
```

>   {icon="x"}

Reviewer: - Support an `icon-default` attribute. All icons in the list will take up the attributes here (icon, icon-class/width/height/size/etc.)
- `icon-default` is by default the first list item's attributes

Did I get this right? It sounds good to me.

@damithc

>   {icon="x"}

Reviewer 2: > Did I get this right? It sounds good to me.

Yes, that sounds good.

Separately, does the `icon-default` of a nested sub-list apply to that sub-list only, or all nested sub-lists at that level?

>   {icon="x"}

Author: it should apply to that sub-list only otherwise the logic for a list will be very messy:
for example:
```
* list 1 
  * list 1.1 (icon from 5.1) // User will be having a hard time locate for a long complex list
  * list 1.2 (icon from 5.1) //
* list 2 
  * list 2.1
  * list 2.2
* list 3
* list 4
* list 5
  * list 5.1 (default)

```

>   {icon="x"}

Author: > For the example above, why would the user put the default at the bottom of the list? The more intuitive thing is to specify it as higher in the list as possible (in this case, item 2.1), right?

I don't think place the default at 2.1 or 1.1 or 5.1 matters, the key ideas here is user has to balling where is the default attribute in the list is place at.
To me , it is my personal bias that i believe md should be as straight forward as it is 😂
> That said, I'm fine to scope it to the sub list if the implementation is too complex for the alternative. Particularly so as this is the first version of the feature. It is fine to go with a MVP first.

I understand the concern about users having to specify multiple defaults, which could lead to code duplication. However, much like the previous point, I believe the Markdown (md) should remain as straightforward as possible. The implementation wise do not worry me.

One alternative approach could be to inherit from the parent. So, instead of using the default md dot and white dot, we would populate the list with the default icon, if specified. Each default icon in a nested layer would then take priority over its parent. At the very least, this approach maintains a tree relationship for the default icons.

>   {icon="x"}

Author: Base on the suggestion above we will encounter something like this:
```
* item 1 {icon="this is the exception"}
  * item 1 {icon="this is the exception"} 
  * item 1 {icon="this is the exception"} 
  * item 1 {icon="this is the exception"}
* item 2 {icon="this is the default"}  
```

then user has to specific:

```
* item 1 {icon="this is the exception"}
  * item 1 {icon="this is the default" default}  
  * item 1 {icon="this is the default"}  
  * item 1 {icon="this is the default"}  
* item 2 {icon="this is the default"}  
```

I think this should be fine.

>   {icon="x"}

Reviewer 2: Perhaps the reason why we have this disparity of view is because I view the following _as a list with two levels_, not as _a list with three sub-lists_.  

```
* list 1
  * list 1.1
  * list 1.2
* list 2 
  * list 2.1
  * list 2.2
* list 3
* list 4
* list 5
  * list 5.1
  ```
For the use-case of _changing default icons to ❓  and ✔️_, I would write:
```
* list 1 {icon="question"}
  * list 1.1 (icon="tick"}
  * list 1.2
* list 2 
  * list 2.1
  * list 2.2
* list 3
* list 4
* list 5
  * list 5.1
  ```
output:
```
❓  list 1
  ✔️ list 1.1
  ✔️ list 1.2
❓  list 2 
  ✔️ list 2.1
  ✔️ list 2.2
❓  list 3
❓  list 4
❓  list 5
  ✔️ list 5.1
 ```
Note: There is no need for `default-icon` as the first icon specification at each level is taken as default.

If there is a need to use specific icons for some items (e.g., use ❌ to indicate wrong answer):
```
* list 1 {icon="question"}
  * list 1.1 (icon="x"}
  * list 1.2 {icon="tick" default-icon}
* list 2 
  * list 2.1
  * list 2.2 (icon="x"}
* list 3
* list 4
* list 5
  * list 5.1
```
output:
```
❓  list 1
  ❌ list 1.1
  ✔️ list 1.2
❓  list 2 
  ✔️ list 2.1
  ❌list 2.2
❓  list 3
❓  list 4
❓  list 5
  ✔️ list 5.1
 ```
To me, the above syntax is straight forward to use and understand. But this kind of thing are subjective as it depends on each person's way of viewing things. So, I will not insist on the above. I'm fine if the feature is slightly more verbose for the above cases (i.e., if we go with scoping for each sub level).

>   {icon="x"}

Author: > output:
> 
> ```
> ❓  list 1
>   ❌ list 1.1
>   ✔️ list 1.2
> ❓  list 2 
>   ✔️ list 2.1
>   ❌list 2.2
> ❓  list 3
> ❓  list 4
> ❓  list 5
>   ✔️ list 5.1
> ```
> 
> To me, the above syntax is straight forward to use and understand. But this kind of thing are subjective as it depends on each person's way of viewing things. So, I will not insist on the above. I'm fine if the feature is slightly more verbose for the above cases (i.e., if we go with scoping for each sub level).

The waterfall model is quite effective, aside from the issue mentioned at: [Comment](https://github.com/MarkBind/markbind/pull/2316#discussion_r1253109949).

There might be a typo here
```
> ```
> ❓  list 1
>   ❌ list 1.1
>   ✔️ list 1.2
> ❓  list 2 
>   ✔️ list 2.1
>   ❌list 2.2
> ❓  list 3
> ❓  list 4
> ❓  list 5
>   ❌ list 5.1
> ```
```

>   {icon="x"}

Reviewer 2: > The waterfall model is quite effective, aside from the issue mentioned at: [Comment](https://github.com/MarkBind/markbind/pull/2316#discussion_r1253109949).

I'm not advocating the waterfall model -- this is the same specify-any-item-as-default-and-override-where-needed model you proposed, which I think is a good approach. The user is free to put the `default-icon` in any item, but the top of the list is the natural place to put it (as we read from top to bottom), so I expect most users will do that. 

Secondly, I'm advocating that we treat all level-2 items together rather than separate sub-lists (i.e., only one `default-icon` for all level-2 items). But as I mentioned, I will not insist on it as it is a matter for personal preference.

>   {icon="x"}

Reviewer: > One alternative approach could be to inherit from the parent. So, instead of using the default md dot and white dot, we would populate the list with the default icon, if specified. Each default icon in a nested layer would then take priority over its parent. At the very least, this approach maintains a tree relationship for the default icons.

Imo its too counterintuitive versus picking one of the below.

> From a user POV, replacing the regular icon of level n with a different icon for the entire list would be a common use case e.g..

> I understand the concern about users having to specify multiple defaults, which could lead to code duplication. However, much like the previous point, I believe the md should remain as straightforward as possible. The implementation wise do not worry me.

Tough one... This would be a really good candidate for a/b testing but we don't have that luxury 😿

Adding a few other considerations (but I'm fine with trying any and pivoting later as well if needed)

1. How often would we expect sublists at the same level to be broken up?
2. How often can we expect extreme cases like:
   ```
   - a
      - a {icon="i-need-to-be-repeated-100-times"}
   - a
      - a {icon="i-need-to-be-repeated-100-times"}
   ... times 100
   ```
   ? (e.g. large student lists, lists of book authors, ...)

   More troublesome but you could also argue for a nunjucks macro in this case..

>   {icon="x"}

Reviewer 2: > I actually like the waterfall logic, as it is very intuitive, combined with the defalt attribute can greatly enhance the UX.

I think so too.

> There are circumstances when users want to use the default icon for different nested lists, but at some levels, there may be conflicts. How could we resolve these?

Related to another question I raised earlier. Assuming we allow specifying `default-icon` multiple times, if the user wants to change the default icon at any point, just use another `default-icon` at that point. Would that work?

>   {icon="x"}

Author: > {icon="i-need-to-be-repeated-100-times"}

If we use nunjucks then all these issue will not be an issue, because it is programable, user do not need to hard code. and everything will be nicely handled by backend.

> How would this work?
> 
> ```
> - a {icon="i-am-the-icon-for-items-1-till-3-because-of-waterfall"}
> - a
> - a
> - a {icon="i-am-the-icon-for-items-4-till-5-because-of-waterfall" default-icon}
> - a
> ```
> Not sure how `default-icon` would have any impact

This is a pure waterfall logic, the default-icon here is redundant.
The conflict lies between the "default" attribute and the idea of treating all "level-2 items" together, as we permit the default attribute to be placed anywhere the user desires.

Both concepts have limitations. Perhaps try the all level waterfall logic, as it uses least attributes?

>   {icon="x"}

Reviewer 2: > The conflict lies between the `default` scope list level attribute and the idea of treating all `level-2 items` together, as we permit the `default` attribute to be placed anywhere the user desires.

Yup.

>   {icon="x"}

Reviewer: Ah... waterfalling the default across sublists (correct me if I'm wrong)

>   {icon="x"}

Reviewer: > Both concepts have limitations. Perhaps try the all level waterfall logic, as it uses least attributes?

Sounds good to me.

To make sure we're on the same page, what would happen in these cases?

```
- a {icon="icon-1"}
- a
- a
- b {icon="icon-2"}
- b
```

```
- a {icon="icon-1"}
  - a {icon="icon-3"}
  - a
  - a {icon="icon-4"}
- a
- b {icon="icon-2"}
  - a
  - a {icon="icon-5"}
  - a
- b
```

```
- a
  - a
  - a {icon="icon-1"}
- b
  - b
- c {icon="icon-2"}
```

>   {icon="x"}

Reviewer 2: My take:
```
- 1 a {icon="icon-1"}
- 1 a
- 1 a
- 2 b {icon="icon-2"}
- 1 b
 ```

```
- 1 a {icon="icon-1"}
  - 3 a {icon="icon-3"}
  - 3 a
  - 4 a {icon="icon-4"}
- 1 a
- 2 b {icon="icon-2"}
  - 3 a
  - 5 a {icon="icon-5"}
  - 3 a
- 1 b
```
```
- a
  - a
  - 1 a {icon="icon-1"}
- b
  - b
- 2 c {icon="icon-2"}
```

>   {icon="x"}

Author: > > Both concepts have limitations. Perhaps try the all level waterfall logic, as it uses least attributes?
> 
> Sounds good to me.
> 
> To make sure we're on the same page, what would happen in these cases?
> 
> ```
> - a {icon="icon-1"}
> - a 1
> - a 1
> - b {icon="icon-2"}
> - b 2
> ```
> 
> ```
> - a {icon="icon-1"}
>   - a {icon="icon-3"}
>   - a 3
>   - a {icon="icon-4"}
> - a 1
> - b {icon="icon-2"}
>   - a 2
>   - a {icon="icon-5"}
>   - a 5
> - b 2
> ```
> 
> ```
> - a 2
>   - a 5
>   - a {icon="icon-1"}
> - b 2
>   - b 1
> - c {icon="icon-2"}
> ```

> To make sure we're on the same page, what would happen in these cases?

This is a good example 👍. 

Currently, "first and default" is the list's implementation, and we just need to expand this logic to all list levels instead of just at the scope level.

I'm good with both `first and default` and `waterfall` logic.

Should we confirm which one?

>   {icon="x"}

Reviewer 2: @SPWwj your answer is very different from mine. 

So my proposal (which is `default + waterfall` ) is as follows:

1. If an item has an icon specified in it, that icon will be used for it. 
1. If the `default-icon` keyword is in the icon specification, that icon will be used as the default for the subsequent items in that level.
1. If an item has no icon specified, the current default icon for that level will be used. If there is no default icon for that level yet, the standard bullet points will be used.
1. The default icon for a level may be changed at any point of the list by using the `default-icon` keyword again. The new default will apply to subsequent items in that level.

I omitted the rule about the icon in the first item becoming the `default-icon` automatically. I think it is cleaner to insist on the `default-icon` keyword, although slightly more work for the user.
I assumed we allow standard bullets to be mixed with custom icons. Not sure if that is feasible. The alternative is 'if one icon is customized, all icons must be customized'. 
As per waterfall strategy, the `default-icon` cannot flow upwards. It applies to items below it only.
The  last rule is already implied by the other rules, but I mentioned it for clarify. 

What do you think?

>   {icon="x"}

Author: > @SPWwj your answer is very different from mine.
> 
> So my proposal (which is `default + waterfall` ) is as follows:
> 
> 1. If an item has an icon specified in it, that icon will be used for it.
> 2. If the `default-icon` keyword is in the icon specification, that icon will be used as the default for the subsequent items in that level.
> 3. If an item has no icon specified, the current default icon for that level will be used. If there is no default icon for that level yet, the standard bullet points will be used.
> 4. The default icon for a level may be changed at any point of the list by using the `default-icon` keyword again. The new default will apply to subsequent items in that level.

> Not sure if that is feasible. The alternative is 'if one icon is customized, all icons must be customized'. As per waterfall strategy, the `default-icon` cannot flow upwards. It applies to items below it only. The last rule is already implied by the other rules, but I mentioned it for clarify.
> 
> What do you think?

It is doable, but mixing the standard bullests with customs icon may not look appeal because they have different line height from the icon, and we do not want to effect the behavier of original markdown bullet by setting a fix line height for all items.
<img width="92" alt="image" src="https://github.com/MarkBind/markbind/assets/30100720/74ee132a-8d04-49e9-9654-31a7c3b208cb">

If we want to apply the waterfall logic to all lists at the same level, here's a suggestion:

 1. If an item has an icon specified in it, that icon will be used for it.
 2. If the `default-icon` keyword is in the icon specification, that icon will be used as the default for the subsequent items in that level.
 3. If an item has no icon specified, the current default icon for that level will be used. If there is no default icon for that level yet,**The nearest icon from the top will be used.** otherwise the **The nearest icon from the bottom will be used.**
 5. The default icon for a level may be changed at any point of the list by using the `default-icon` keyword again. The new default will apply to subsequent items in that level.

The augmented point 3 has actually diminished the importance of the "default" attribute.

Alternatively, we can employ the "first and/or default" approach as demonstrated in your output:
```
- 1 a {icon="icon-1"}
  - 3 a {icon="icon-3"}
  - 3 a
  - 4 a {icon="icon-4"}
- 1 a
- 2 b {icon="icon-2"}
  - 3 a
  - 5 a {icon="icon-5"}
  - 3 a
- 1 b
```

With this approach, users can place "default" at any item within the same level. If not specified, the first icon automatically becomes the default.

>   {icon="x"}

Reviewer 2: > the nearest icon from the top will be used, otherwise the The nearest icon from the bottom will be used.

While this works, it is not intuitive/natural. The nearest icon may have been intended as an one-off override. 

On a related note, I feel it would be easier for users to understand (and for us to explain to users) if all defaults are explicitly marked as `default-icon` (and only apply downwards) and all icons without `default-icon` are treated as overrides (and apply to that item only). 

> but mixing the standard bullests with customs icon may not look appeal because they have different line height from the icon, and we do not want to affect the behavier of original markdown bullet by setting a fix line height for all items.

I agree. So, I don't expect users to define lists this way, but we can't leave this case 'undefined' either. If they mix custom icons and standard icons, the list will look ugly, but that's their choice.

>   {icon="x"}

Author: What if we introduce a default icon as a placeholder for undefined cases? This icon would match the size of the nearest icon in the list, by doing this, we maintain the consistency of undefined cases.

>   {icon="x"}

Author: > substantial

It look quite well, even with different sizes of icons.

The current version takes the first occurrence as the default. We can use this feature to test and assess the user experience.

<img width="332" alt="image" src="https://github.com/MarkBind/markbind/assets/30100720/a2bfcc61-83ed-4e35-9cad-e253d26aaecd">

```
- item 1
- item 2 {size="80px" icon="https://raw.githubusercontent.com/SPWwj/Spirify-Ui/c26edc9581162663d29f4d8f911f98fb7f971ca6/src/assets/svgs/image-error.svg"}
- item 2 { size="80px" icon="https://github.com/SPWwj/Spirify-Ui/blob/master/src/assets/images/event.png?raw=true"}
- item 3 { size="25px" icon="https://github.com/SPWwj/Spirify-Ui/blob/master/src/assets/images/sharing.png?raw=true"}
- item 4 { size="120px" icon="https://github.com/SPWwj/Spirify-Ui/blob/master/src/assets/images/event.png?raw=true"}
- item 5 {icon="https://github.com/SPWwj/Spirify-Ui/blob/master/src/assets/images/attraction.png?raw=true"}
```
While it's feasible to mix standard bullets with customized icons, I wouldn't recommend it. Not only does it potentially compromise aesthetics, but it also introduces complexity due to the icons belonging to different data structures. Any updates made to related components could impact each other, similar to the case when MarkBind removed the 'margin-top' for the 'p' tag, making it challenging to distinguish between loose and tight lists in some scenarios.

>   {icon="x"}

Reviewer 2: > It look quite well, even with different sizes of icons.
> 
> The current version takes the first occurrence as the default. We can use this feature to test and assess the user experience.
> 
> <img alt="image" width="332" src="https://user-images.githubusercontent.com/30100720/251745925-a2bfcc61-83ed-4e35-9cad-e253d26aaecd.png">

What's your default placeholder icon for level 1, 2, 3? We should not pick any random images as default placeholders. They should be a close match to standard bullet points. e.g., fontawsome equivalent for black dot, white dot, and black square
* default placeholder for level 1
  * default placeholder for level 2
    * default placeholder for level 3

> While it's feasible to mix standard bullets with customized icons, I wouldn't recommend it.

In that case we can use fontawsome equivalent for standard bullets

>   {icon="x"}

Author: I hadn't considered that far. As you've noticed, it's merely an error SVG image for an undefined icon. We can certainly determine placeholder images for levels 1, 2, and 3. Since it's an icon list, we have the freedom to decide its appearance, preferably keeping it as close as possible to the standard bullet points.

<img width="402" alt="image" src="https://github.com/MarkBind/markbind/assets/30100720/095b17c2-3e86-43bb-ae87-e02e5f760f5b">

>   {icon="x"}

Author: <img width="203" alt="image" src="https://github.com/MarkBind/markbind/assets/30100720/97f04f9c-ca97-4cdd-a35c-697f8784459b">

>   {icon="x"}

Author: <img width="419" alt="image" src="https://github.com/MarkBind/markbind/assets/30100720/7032d22e-a7ba-4a2f-9261-9b4c028fcc02">

We can align the text to the center, ensuring it looks appealing in all situations.

>   {icon="x"}

Reviewer 2: This is very unlikely what the user wanted, or would ever want. I'd rather use standard bullets, or if that is not feasible, something that looks like standard bullet points (appearance, size, color etc.) **so that user can clearly see they remain un-customized**. We need not try to auto-customize them because this is not meant to be the final form of them anyway.

>   {icon="x"}

Author: Is this one better

<img width="184" alt="image" src="https://github.com/MarkBind/markbind/assets/30100720/b4fe541e-7d05-4357-a1a5-dc012b8706b8">

Or we go for this one:

![image](https://github.com/MarkBind/markbind/assets/30100720/95ae0f74-af33-43d8-8da1-6430f4858b72)

>   {icon="x"}

Reviewer 2: @SPWwj  I prefer the first one (i.e., standard bullets) but you mentioned it has some extra implementation costs, in which case the second one (i.e., similar looking icons from elsewhere) is fine too.

>   {icon="x"}

Reviewer: > I think users will realize by themselves that it is not a good idea to mix customized icons with standard icons, without us trying to second-guess them.

@damithc @SPWwj how about simply:

1. > but we can't leave this case 'undefined' either 

   Fail hard if we see the user trying to mix the two:
   - waterfall: fail hard if we see the user trying to mix the two in any single level of lists
   - first + default: fail hard if there is no default icon defined.

   - "Fail hard" can be: logging a warning + failing the build
2. If there's really a need for mixing the two, you'd still be able to specify something that looks close to the default from fontawesome/octicon/one of our supported libraries.

>   {icon="x"}

Reviewer: > default + waterfall

My interpretation of "waterfall" is the [original](https://github.com/MarkBind/markbind/pull/2316#discussion_r1252062287) one with no `default-icon` attribute at all, based on @SPWwj's wording here:

> Both concepts have limitations. Perhaps try the all level waterfall logic, as it uses **least attributes?**

Personally I'd prefer to have no `default-icon` attribute at all if we are going with the waterfall model (otherwise it should be named something like `default-waterfall-icon`).

---

Too recap things too as this thread is getting very long...

- Pure waterfall
   - "Icon shall be X until I see a new one at the same level"
   - design can be more verbose in some cases
      - https://github.com/MarkBind/markbind/pull/2316#discussion_r1253077627 (and comment below that) 
   - for the multiple sublists at the same level issue, it'll be handled by design
- `default-icon`, first `icon` is the default.
   - First `icon` is the default. If its the exception, the author can recontrol it using `default-icon`.
   - for the multiple sublists at the same level issue, we have 2 choices:
      - treat all sublists as one
          - Con: more verbose if sublists should have different defaults
      - treat them separately, defaults should be respecified for every list
	      - I'd prefer this personally between the two options of treating sublists, its simpler to reason with as @SPWwj has mentioned
          - Con: more verbose if all sublists should have the same defaults

Both approaches should be adequate to support simple and common use cases conveniently. For more complex cases, it should not be impossible either, only slightly more verbose.

>   {icon="x"}

Reviewer: > Too recap things too as this thread is getting very long...

Overall my personal choice would be `first + default + treat sublists separately`.

- Waterfall is convenient but not the most conventional design.
- In the scenario we need to configure different sublists at the same level separately, things are far easier than the other 2 approaches. It is a rarer case, but I don't think we should be making rare cases too difficult to configure either.
   - e.g. sublist A should have icon X but sublist B should have no icon (in this case I realise we'd actually need to support an additional `no-icon` attribute of sorts for the other 2 approaches as well)
   -  e.g. sublist A should have icon X but sublist B should have icon Y. (as easy as waterfall, but harder for `first + default + treat sublists together`)
- It is more troublesome when all sublists share the same icon but not by too much.
   - for small lists (<10), a few ctrl-c + ctrl-v's
   - for larger lists, you'd likely be using nunjucks loops already, and there is nunjucks macros/variables

>   {icon="x"}

Author: > * convenient

Both demonstrations use FontAwesome😂. We can customize the icons to look as close as possible, or even identical, to standard bullet points.

>   {icon="x"}

Author: > Overall my personal choice would be first + default + treat sublists separately.
Waterfall is convenient but not the most conventional design.
In the scenario we need to configure different sublists at the same level separately, things are far easier than the other 2 approaches. It is a rarer case, but I don't think we should be making rare cases too difficult to configure either.
e.g. sublist A should have icon X but sublist B should have no icon use the default browser icon (in this case I realise we'd actually need to support an additional no-icon use-browser-default-icon attribute of sorts or icon="use-browser-default" for the other 2 approaches as well)
e.g. sublist A should have icon X but sublist B should have icon Y. (as easy as waterfall, but harder for first + default + treat sublists together)
It is more troublesome when all sublists share the same icon but not by too much.
for small lists (<10), a few ctrl-c + ctrl-v's
for larger lists, you'd likely be using nunjucks loops already, and there is nunjucks macros/variables

Thanks for the recap. To add to this, I believe the choice between 'waterfall' and 'first or default' should be based on the use case.

`Waterfal` would be advantageous when there are many sublists of the same level using the same default icon.
  - Use cases: to-do lists, checklists, progress lists.

`First or default`, on the other hand, is suitable when different sublists need to use different icons.
  - Use cases: team lists, domain-specific lists like subject lists where each subject has different default icons and focus areas.

Lastly, in the later stages if needed, we can certainly incorporate JavaScript into the icons to enhance their functionality. This could potentially replace existing Task lists (from GitHub Flavored Markdown) or Radio-button lists.

>   {icon="x"}

Reviewer 2: I think we have different interpretation of waterfall etc. so let's focus on how we are going to explain this to users, as that will be the final delivery format. 

Shall we go with the following then? Not the most efficient in terms of how much extra code user needs to write, but easiest to explain.

1. If an item has an icon specified in it, that icon will be used for it, and subsequent items at that level.
2. If any item in a level is customized, the first item in that level must be customized as well. Otherwise the list will be rejected.

We can replace `level` with `sub-list`, if we want to treat each sub list separately.

>   {icon="x"}

Author: If our goal is to clarify, then 'first or default' would be the most straightforward and intuitive option. Here's how it works:

- If an item has an icon specified, that icon will be used.
- If an item doesn't have an icon, the default icon will be used.
- The default icon is determined by the 'default' attributes, or by the first item in the list if and only if the 'default' attributes are missing.
- If multiple defaults are present, the first default will be given priority.

> 1. If an item has an icon specified in it, that icon will be used for it, and subsequent items at that level.
> 1. If any item in a level is customized, the first item in that level must be customized as well. Otherwise the list will be rejected.

Regarding the second point, from a user's perspective, this could generate a lot of questions. For instance, what if I have a long list of, say, 10 items, and I only know the icons for items 3, 6, and 9? Does that mean I need to configure items 1, 3, 4, 6, 7, 9, and 10? That doesn't seem very efficient.

>   {icon="x"}

Reviewer 2: > Regarding the second point, from a user's perspective, this could generate a lot of questions. For instance, what if I have a long list of, say, 10 items, and I only know the icons for items 3, 6, and 9? Does that mean I need to configure items 1, 3, 4, 6, 7, 9, and 10? That doesn't seem very efficient.

I agree, it's not efficient. A default can make it more efficient. But that is something we can add later. This two-rule version is simpler (to explain, and I assume, to implement) and good enough for the MVP. I don't like it's tendency for more verbose code either, but I'm willing to live with that for now.

The user is going from 'not being able to customize lists at all' to 'able to customize every icon', so it is a big step forward for the user despite the verbosity. I don't think any other Markdown based static site generator has that ability. Users have no cause for complains.

Plus, we want to get out of this analysis-paralysis we are stuck in now, and deliver a working MVP first

>   {icon="x"}

Reviewer: I'd still prefer `first and default` for the reasons outlined https://github.com/MarkBind/markbind/pull/2316#discussion_r1257008347. Also, no concept of a placeholder icon (instead, failing hard has the benefit of **_forcing_** the author to format the list properly) (instead of having it leak into a deployment).

> we want to get out of this analysis-paralysis we are stuck in now

I strongly agree with this. I think the choice is between the ones here https://github.com/MarkBind/markbind/pull/2316#discussion_r1256988082.

Let's try to avoid mixing the waterfall + first&default models for now.

>   {icon="x"}

Reviewer 2: The one I proposed last in https://github.com/MarkBind/markbind/pull/2316#discussion_r1257106600 is same as the first one in https://github.com/MarkBind/markbind/pull/2316#discussion_r1256988082 right?

>   {icon="x"}

Reviewer: yup

>   {icon="x"}

Author: > If an item has an icon specified in it, that icon will be used for it, and subsequent items at that level.
> If any item in a level is customized, the first item in that level must be customized as well. Otherwise the list will be rejected.

@ang-zeyu @damithc I've updated the logic to strictly follow a waterfall structure based on the two rules above. Maybe you can try it out and see how it feels. We can adjust and make improvements as necessary.

The checks failed due to an issue with markdown-it-emoji-fixed, but that shouldn't affect other functionalities.

>   {icon="x"}

Reviewer: @SPWwj I'm getting this from building your docs:

<img src="https://github.com/MarkBind/markbind/assets/3306138/a2dd656d-6d30-4cb1-ad07-5506ec330c8a" width=300 >

Otherwise, while I still prefer `first + default + treat sublists separately`, I'm fine with following waterfall as well, its not a strong preference.

I think we've gathered quite a few number of valuable opinions and ideas by this point too. In the interest of time and preventing decision paralysis, I suggest @damithc make the call at this point.

I'll just make some observations on the differences/additions of your current implementation from the 2 rules since the docs is broken:
- Sublists seem to be treated separately. (I hadn't thought of / noticed this previously, I'll add it to the recap list in https://github.com/MarkBind/markbind/pull/2316#discussion_r1256988082 very shortly for comparison (cc @damithc))
   <img src="https://github.com/MarkBind/markbind/assets/3306138/922a12cb-9575-4d2b-862a-ba4eabc8a9cf" width=300 >

- nearest icon is taken if none is specified
   <img src="https://github.com/MarkBind/markbind/assets/3306138/0dea9b17-69a0-4101-8d1b-32b8d99015e4" width=300 >

>   {icon="x"}

Reviewer: > The checks failed due to an issue with markdown-it-emoji-fixed, but that shouldn't affect other functionalities.

You should be able to do `import * as emojiData from '../lib/markdown-it/patches/markdown-it-emoji-fixed'` instead.

>   {icon="x"}

Author: > @SPWwj I'm getting this from building your docs:
> 
> <img alt="" width="300" src="https://user-images.githubusercontent.com/3306138/252020400-a2dd656d-6d30-4cb1-ad07-5506ec330c8a.png">
> Otherwise, while I still prefer `first + default + treat sublists separately`, I'm fine with following waterfall as well, its not a strong preference.
> 
> I think we've gathered quite a few number of valuable opinions and ideas by this point too. In the interest of time and preventing decision paralysis, I suggest @damithc make the call at this point.
> 
> I'll just make some observations on the differences/additions of your current implementation from the 2 rules since the docs is broken:
> 
> * Sublists seem to be treated separately. (I hadn't thought of / noticed this previously, I'll add it to the recap list in [Add Icon and Media Support to List  - Resolves #906 and #2272 #2316 (comment)](https://github.com/MarkBind/markbind/pull/2316#discussion_r1256988082) very shortly for comparison (cc @damithc))
>   <img alt="" width="300" src="https://user-images.githubusercontent.com/3306138/252021601-922a12cb-9575-4d2b-862a-ba4eabc8a9cf.png">
> * nearest icon is taken if none is specified
>   <img alt="" width="300" src="https://user-images.githubusercontent.com/3306138/252021983-0dea9b17-69a0-4101-8d1b-32b8d99015e4.png">

Did you update the branch to the latest [2548b86](https://github.com/MarkBind/markbind/pull/2316/commits/2548b862929e68e3f3af8aaa8b20d7bdbfd2264a)?

As demonstrated by the screenshots in the PR description, it works fine for me. The two rules are also strictly followed.

<img width="553" alt="image" src="https://github.com/MarkBind/markbind/assets/30100720/3b4ea552-bde0-4355-805d-fea3df85b420">

I will be back tmr 😂

>   {icon="x"}

Reviewer: @SPWwj  yes, its the wrong build.

Let me know too if I missed anything:

> `<insert if anything missing>` @SPWwj

>   {icon="x"}

Author: @ang-zeyu I know why, there is a migration for markdown-it-emoji-fixed.js to ts needed, the npm run clean will remove my markdown-it-emoji-fixed.d.ts file. 

https://github.com/MarkBind/markbind/blob/2548b862929e68e3f3af8aaa8b20d7bdbfd2264a/packages/core/src/lib/markdown-it/patches/markdown-it-emoji-fixed.d.ts

>   {icon="x"}

Reviewer: It only removes those that have a corresponding `'**/*.js.map'` file (which comes from tsc compiling `.ts`).

You'll need to either `import * as emojiData from '../lib/markdown-it/patches/markdown-it-emoji-fixed'` instead ([ts docs](https://www.typescriptlang.org/docs/handbook/modules.html#import-the-entire-module-into-a-single-variable-and-use-it-to-access-the-module-exports)) or modify the export `module.exports = { content: emojiData }`

>   {icon="x"}

Author: The d.ts file is an auto generated typing file for ts thus will get removed, this is the reason why the build only works for my local machines😂 The method you mentioned above isn’t what is been used currently? Just that due to tsconfig importing untyped file will throw an error(thus I created d.ts file), I will get back to this soon…

>   {icon="x"}

Reviewer: > The d.ts file is an auto generated typing file for ts thus will get removed, this is the reason why the build only works for my local machines😂

@SPWwj your `.d.ts` file was manually created so it will be skipped in the `npm run clean` script, that is not the issue. You are attempting to do an es6 named import from a CJS module that exports `module.exports = emojiData` which makes no sense (see above 2 solutions). Typescript definition files do not affect the compilation output here but only type checking.

> The method you mentioned above isn’t what is been used currently?

Please refer to the ts docs ([1](https://www.typescriptlang.org/docs/handbook/modules.html#import-the-entire-module-into-a-single-variable-and-use-it-to-access-the-module-exports), [2](https://www.typescriptlang.org/docs/handbook/modules.html#export--and-import--require)). We have a typescript [migration guide](https://markbind.org/devdocs/devGuide/development/migratingToTypeScript.html#import-export-syntax-reference) in the dg as well. 

> thus I created d.ts file

It's good you've created a definition file for type safety. If you'd like to proceed without it, you'll need to do a regular `const emojiData = require('../lib/markdown-it/patches/markdown-it-emoji-fixed');` so that tsc skips type checking it.

Let me know if its still not clear.

>   {icon="x"}

Reviewer: https://github.com/MarkBind/markbind/pull/2316#discussion_r1257313936

@damithc based on your review, I assume you are fine with the current approach?

>   {icon="x"}

Reviewer 2: > @damithc based on your review, I assume you are fine with the current approach?

Sorry, I missed this one. @SPWwj I wasn't following this thread as it was about the implementation details. What's the 'current approach' referred to here?

>   {icon="x"}

Author: > Sorry, I missed this one. @SPWwj I wasn't following this thread as it was about the implementation details. What's the 'current approach' referred to here?

Haha, it refers to the first and default model. The waterfall is just another layer on top of it. However, the current implementation is based on the waterfall model, since there hasn't been any objection to it.

>   {icon="x"}

Reviewer: pinging again @damithc 

@SPWwj If this is ok we can start writing some tests too.

>   {icon="x"}

Reviewer 2: > pinging again @damithc

Sorry @ang-zeyu , I'm not sure what's the thing I need to comment on. I'm OK with the current waterfall approach (as per the description in the UG).

* item 3
  
{icon="fas fa-check"}
</variable>
</include>

</div>

<box type="tip" seamless>

> The icon appearance can be customized further a class attribute.

Reviewer 2: typo

</box>
<div id="main-example-markbind">

<include src="codeAndOutput.md" boilerplate >
<variable name="highlightStyle">markdown</variable>
<variable name="code">
* item 1 {icon="fas fa-check" class="text-warning"}
* item 2 {icon="fas fa-check" class="text-danger"}
* item 3 
  
{icon="fas fa-check" class="text-primary"}
</variable>
</include>

> </div>

Reviewer 2: Note that icons (just like characters) have different widths, even if the size is the same. However, we want all items/icons in the list to be aligned nicely, if the icons are of same size. Otherwise the list will look very untidy (like the example below -- I used letters instead of icons).

**X**  item 1
**I**  item 2
**W**  item 3 

When aligned properly:
`X`  item 1
`I`  item 2
`W`  item 3

Author: The following has been taken into account:
 
<img width="437" alt="image" src="https://github.com/MarkBind/markbind/assets/30100720/cb72639e-c067-4ed2-88be-5000822ee713">

Reviewer 2: I see. Should be fine if you also considered that the widths of characters differ even at the same font size (same for font-based icons such as fontawesome).
Anyway, you can test with something like this, to see if the outcome is aligned nicely:
```
* item 1 { icon="fas home" } 
* item 1 { icon="fas trophy" } 
* item 1 { icon="glyphicon education" } 
 ```

Author: Here is the sample output:
<img width="503" alt="image" src="https://github.com/MarkBind/markbind/assets/30100720/1f3197b3-0b68-48f0-8e63-9467f71ed785">

Reviewer 2: Looks good. Seems that problem has been taken care of already 👍

Reviewer: Imo, fixed sizes are user friendly but I often find myself wanting to do more fine grained adjustments and it ends up being less so.

How about:
- Icon used: `height=width` by default (as icons tend to be square), `width=1em` (responsive to text size)
- Image used: `height=50px` (or anything else that makes sense), let `width` adjust according to the image.
- Support specifiying any units in `icon-width + icon-height` attributes 

In consideration with the below comment on aspect ratios too*

Reviewer: https://github.com/MarkBind/markbind/pull/2316#discussion_r1252088583

Reviewer: Another thing if you have the time, you can consider is providing an alignment option for the case that icon widths are different (e.g. images of different sizes); But I think this is rarer and less important for now, we can pick a good default and add it later if needed.

Author: Imo excessive freedom doesn't always yield positive results. Although an image can have a 16:9 ratio or non-square dimensions, it's easy to envision that such an image wouldn't work well as an icon and may not be aesthetically pleasing. Nonetheless, it's still worth considering.

Reviewer: if you mean aesthetics, it can be fairly pleasing:

![image](https://github.com/MarkBind/markbind/assets/3306138/4b92a131-2706-4a47-9aa1-2bfecf9c8e21)

it'd be up to the author to use something appropriate, we merely provide that flexibility of non-square images.

Reviewer: > Image used: width=50px (or anything else that makes sense), let height adjust according to the image.

If the author's original image were already square, it should also work out nicely without extra configuration.

Reviewer: > We can indeed let users define aspect ratio

Just so we're on the same page I'm not suggesting an `icon-aspect-ratio` of sorts, instead authors should be able to let the aspect ratio be defined by their media size (while being able to specify height, or width).

Similar to how you would configure a regular `<img>`

> Restrict setting specific width and height aims to prevent unpredictable displays, such as a 500px vs 30px scenario.

Not sure I understood this, could you clarify?

> Also, we need to consider mobile view display.

I don't think this is something we can do anything much about in either case. Unless we do something drastic like placing the icon above the content which seems like a bad idea.

Again, similar to how you would think of putting a regular `<img>` on mobile, it is up to the author to size it appropriately.

Some minor "solutions" I can think of currently otherwise:
- drop a well placed documentation note/tip on this.
- Maybe also tighten our existing margins/paddings.
- something like `icon-mobile-size/icon-mobile-width` but I'm not sure if its necessary for now, we can address it if the need pops up. (considering you already added a `icon-class` feature too)

Author: > Just so we're on the same page I'm not suggesting an `icon-aspect-ratio` of sorts, instead authors should be able to let the aspect ratio be defined by their media size (while being able to specify height, or width).

This is just my design consideration.

> Not sure I understood this, could you clarify?

For instance, consider iPhone widgets. Developers can only create widgets of specific sizes. This restriction enables users to have an optimal display, irrespective of the size and device they use.

> * something like `icon-mobile-size/icon-mobile-width` but I'm not sure if its necessary for now, we can address it if the need pops up. (considering you already added a `icon-class` feature too)

I don't think that 'icon-mobile-size' or 'icon-mobile-width' are good suggestions. Imo, all these tedious specifications should be handled by the framework to optimize UX.
The width and height should be the most lenient we can go. They should be sufficient for our use case.

Reviewer: Sorry, I'm still not quite sure what you are suggesting (if any). To clarify, are you good with this approach as well? https://github.com/MarkBind/markbind/pull/2316#discussion_r1252094010

Or are you suggesting we keep the existing square approach, or something else entirely.

> The width and height should be the most lenient we can go. They should be sufficient for our use case.

Yes, it should be.

> Imo, all these tedious specifications should be handled by the framework to optimize UX.

Do you mean we should go even further with something like this?

1. Allow the author to only control `icon-width=Xpx`

1. Don't trust the author to do the right thing:

```css
@media only screen and (max-width: 768px) {
  .icon {
    /* For mobile, it should not be too big nor small */
    min-width: 20px;
    max-width: 100px;
  }
}
```

Reviewer: > Imo this is what we should aim for, as md should be easy to use with minimal consideration.

I agree this should take priority over transparency.

> Do you mean we should go even further with something like this?

What do you think of this? (as a whole) 

- Configurable `icon-width` in all cases.
- Icon used: `height=width` mandatorily, `width=some default`
- Images/svg used: aspect ratio will be mandatorily preserved.
- `max-width: 100px` (feel free to adjust the number) on some mobile breakpoint
   - `min-width` may not be necessary as the author will notice the problem even on non-mobile devices

If we have a need for `icon-height` down the line we can add it

Author: Here is my suggestions:
Size: square (default to minimal if no size specified),
Optional: width (overrides default),
Optional: height (overrides default).

"Max-width/height: 100px" is a redundant attribute. If users are mindful about this, display issues won't be a concern. However, we can still include it as an optional attribute.

Reviewer: > Size: square

I agree on icons but strongly think images/svg/media should use their native aspect ratio by default. If the image was already meant to be square - it'll stay that way. Rectangular/portrait images should stay that way as well.

> "Max-width/height: 100px" is a redundant attribute. If users are mindful about this, display issues won't be a concern. However, we can still include it as an optional attribute.

Dosen't this go against ⬇️?

> > Don't trust the author to do the right thing:
>
> Imo this is what we should aim for, as md should be easy to use with minimal consideration.

Reviewer: > I agree on icons but strongly think images/svg/media should use their native aspect ratio by default. If the image was already meant to be square - it'll stay that way. Rectangular/portrait images should stay that way as well.

If square media icons are preferred - the author should crop it appropriately using some external software to preserve an unskewed image. Think the circular crop UI you see around profile pages - though we don't have that luxury in markdown (imo it'd be too difficult to use/implement in markdown)

Reviewer: > "Max-width/height: 100px" is a redundant attribute

Not as an attribute to be clear, but the CSS we talked about earlier:

> > Don't trust the author to do the right thing:
> > ```css
> > @media only screen and (max-width: 768px) {
> >   .icon {
> >     /* For mobile, it should not be too big nor small */
> >     min-width: 20px;
> >     max-width: 100px;
> >   }
> > }
> > ```
>
> Imo this is what we should aim for, as md should be easy to use with minimal consideration.

Author: > > Imo this is what we should aim for, as md should be easy to use with minimal consideration.

I see your point, but it might confuse users when their height/width settings don't work, because they didn't specify the min or max attributes.

Reviewer: Not as a HTML attribute at all; But CSS.

> Imo this is what we should aim for, as md should be easy to use with minimal consideration.

Isn't this your point? *quite confused* 😂

Yes, I've mentioned transparency earlier. On second thought (with your point on easy to use), not trusting the author should take priority.

The breakpoint dosen't have to be `768px` as well, something even smaller like `576px` could work. Similarly, the `max-width` dosen't  have to be `100px`, maybe something like `200px`. At the point it would fill nearly half the screen or more (depending on the preexisting indentation of the list) and we can quite confidently question the author's intention).

Author: > Isn't this your point? _quite confused_ 😂

In my opinion, we shouldn't make users feel restricted or surprised by hidden boundaries like a maximum width. Setting up sizes such as XS, S, M, L, XL, instead of width and height with an invisible maximum, could meet user requirements more intuitively. This approach would avoid the frustration of configuring width and height, hitting unexpected limits, and then having to troubleshoot by referring to notes and FAQs in the docs.

Nonetheless, the provided width and height should adequately support the basic usage of all image types within the list.

> The breakpoint dosen't have to be `768px` as well, something even smaller like `576px` could work. Similarly, the `max-width` dosen't have to be `100px`, maybe something like `200px`. At the point it would fill nearly half the screen or more (depending on the preexisting indentation of the list) and we can quite confidently question the author's intention).

However, such an approach could disrupt the list presentation, which may not be ideal. It would be more effective to specify an attribute like {flex} to achieve the desired presentation.

Reviewer: > I am good width width and height suggestions.

I see, so by this you meant we should have `width` + `height` but still stick with fixed sizes to eliminate the possibility for users to construct an inappropriate size on mobile. (correct me if I'm wrong)

Here's my original issue with this approach's UX:

> Imo, fixed sizes are user friendly but I often find myself wanting to do more fine grained adjustments and it ends up being less so.

- `l=70px`
- `xl=96px`

When I'm adjusting media/icons, I often find myself, especially in the case of media, wanting to fine tune things (e.g. to `80px`) to look nice. I end up having to use an unwieldy `style=...` (not even possible here) or define a whole other css class to stick to `icon-class` which is rather inconvenient and counter to having a `size/width/height` attribute.

To fix this I can think of the following approaches:

1. Vuetify for example, https://vuetifyjs.com/en/api/v-icon/#props-size, supports predefined sizes, but also finer grained adjustments with px. I'm fine with following this too if you strongly believe having `xs,s,...` alongside would be worth it.
2. Support only the form that can adequately handle all use cases (specific widths and heights in `px`) for feature simplicity. I'm assuming that someone using a static site generator would have at least a brief idea of what a `px` is, or be able to intuitively bump up or down a number controlling the icon size.

As we discussed there's also an issue here of authors specifying too large a width on mobile and things going haywire. If we would like to "fix" this, we could:

0. Don't fix it at all, trust the author

1. Specifiy `max-width` on mobile devices

   > In my opinion, we shouldn't make users feel restricted or surprised by hidden boundaries like a maximum width.

   Transparency is indeed an issue, but I'd ask a few more questions as this is a tradeoff: Would having a list item icon fill half the screen on mobile device be a reasonable use case? And how much should we value supporting such use cases over authors not controlling widths properly. (imo, something generous like `max-width: 250px` on `576px` breakpoint could work to break this tie)

2. > However, such an approach could disrupt the list presentation, which may not be ideal. It would be more effective to specify an attribute like {flex} to achieve the desired presentation.

   What do you think of https://github.com/MarkBind/markbind/pull/2316#discussion_r1254701627? Is this something like what you have in mind?

Author: > I see, so by this you meant we should have `width` + `height` but still stick with fixed sizes to eliminate the possibility for users to construct an inappropriate size on mobile. (correct me if I'm wrong)

I'm not sure if you prefer having sizes or not - I'm fine with either. All of the above are merely suggestions. The width and height options seem to provide the best flexibility for fine-tuning.

> 1. Vuetify for example, https://vuetifyjs.com/en/api/v-icon/#props-size, supports predefined sizes, but also finer grained adjustments with px. I'm fine with following this too if you strongly believe having `xs,s,...` alongside would be worth it.

I don't particularly advocate for having sizes like 'xs', 's', etc., as shown in your example. It's indeed an interesting approach that other frameworks are implementing. Naturally, we always have the option to define the behavior of the component ourselves, which is one of the wonderful aspects of programming.

> Support only the form that can adequately handle all use cases (specific widths and heights in px) for feature simplicity. I'm assuming that someone using a static site generator would have at least a brief idea of what a px is, or be able to intuitively bump up or down a number controlling the icon size.

That should be quite true.

> As we discussed there's also an issue here of authors specifying too large a width on mobile and things going haywire. If we would like to "fix" this, we could:
> 
> 0. Don't fix it at all, trust the author
> 1. Specifiy `max-width` on mobile devices
>    > In my opinion, we shouldn't make users feel restricted or surprised by hidden boundaries like a maximum width.
>    
>    
>    Transparency is indeed an issue, but I'd ask a few more questions as this is a tradeoff: Would having a list item icon fill half the screen on mobile device be a reasonable use case? And how much should we value supporting such use cases over authors not controlling widths properly. (imo, something generous like `max-width: 250px` on `576px` breakpoint could work to break this tie)
> 2. > However, such an approach could disrupt the list presentation, which may not be ideal. It would be more effective to specify an attribute like {flex} to achieve the desired presentation.
>    
>    
>    What do you think of [Add Icon and Media Support to List  - Resolves #906 and #2272 #2316 (comment)](https://github.com/MarkBind/markbind/pull/2316#discussion_r1254701627)? Is this something like what you have in mind?

I strongly oppose converting the list behavior to a flex view. In my opinion, a list should function as a list unless specified otherwise.

I have done some experiance with different size unit:
```
    width: 50vw;
    height: 50vw;
```
Output:
<img width="658" alt="image" src="https://github.com/MarkBind/markbind/assets/30100720/6eee3804-ec60-4daf-bbfb-768d37e61d63">
<img width="319" alt="image" src="https://github.com/MarkBind/markbind/assets/30100720/f50e2532-7ab7-451f-b46c-7a1b950f4638">

```
width: 400px;
height: 400px;
```
<img width="673" alt="image" src="https://github.com/MarkBind/markbind/assets/30100720/ea61cf83-927b-4705-9591-5b0a7bc548ee">
<img width="456" alt="image" src="https://github.com/MarkBind/markbind/assets/30100720/1403bd85-dc77-47ce-9b10-94faa71a91e4">

As demonstrated by the output, using 'vw' (viewport width) is highly effective in managing the size across different screens.

Perhaps, rather than restricting size to px, we could allow users to utilize any size unit. Whether they use it correctly or not, we can leave that responsibility to them.

Reviewer: > I'm not sure if you prefer having sizes or not - I'm fine with either. All of the above are merely suggestions. The width and height options seem to provide the best flexibility for fine-tuning.

Do you mean having a set fixed sizes? No, I don't prefer having _only_ it as I often find myself wanting to do fine grained adjustments. (see above [example](https://github.com/MarkBind/markbind/pull/2316#discussion_r1255844469)) But having it _alongside_ other css units (`px/vw/etc.`) is fine.

If you mean a single square `size` vs 2 `width/height` attributes, I strongly prefer `width/height` as media should be scaled according to its native aspect ratio to avoid distortion.

---

> I strongly oppose converting the list behavior to a flex view. In my opinion, a list should function as a list unless specified otherwise.

> It would be more effective to specify an attribute like {flex} to achieve the desired presentation.

Did you have something different in mind?

---

Which of all the above approaches do you prefer currently?

https://github.com/MarkBind/markbind/pull/2316#discussion_r1255844469

Personally I'm leaning toward `2` (just css units) and then either `0` (don't enforce max-width at all) or `1` (enforce a _generous_ max-width on mobile).

Author: >Personally I'm leaning toward 2 (just css units) and then either 0 (don't enforce max-width at all) or 1 (enforce a generous max-width on mobile).

Given that there are no strong opinions on this, I suggest we initially go with 0. As demonstrated above, users can choose different size units to accommodate various scenarios and screen sizes.

Reviewer: Sounds good, we can add on `1` later as well if we change our minds.

Reviewer: > > Personally I'm leaning toward 2 (just css units) and then either 0 (don't enforce max-width at all) or 1 (enforce a generous max-width on mobile).
> 
> Given that there are no strong opinions on this, I suggest we initially go with 0. As demonstrated above, users can choose different size units to accommodate various scenarios and screen sizes.

so: `2` + `0` (cmiiw @SPWwj)

Author: Only '0', not '2'. What you suggested is something like a card view rather than a list.

Reviewer: not the later `2` but the earlier one.

> Support only the form that can adequately handle all use cases (specific widths and heights in px) for feature simplicity. I'm assuming that someone using a static site generator would have at least a brief idea of what a px is, or be able to intuitively bump up or down a number controlling the icon size.

Author: >Support only the form that can adequately handle all use cases (specific widths and heights in px) for feature simplicity. I'm assuming that someone using a static site generator would have at least a brief idea of what a px is, or be able to intuitively bump up or down a number controlling the icon size.

Yes, I will go forward with above.

Just to share a trick you can use to achieve what you described - arranging images and text in separate rows on a mobile screen.

 ```
* :glyphicon-education: item 1x {icon=" "}
* ![Cardview](https://github.com/SPWwj/Spirify-Ui/blob/master/src/assets/images/sharing.png?raw=true){width="350px" height="100px"} Hello world
  
  Hello world
```

<img width="444" alt="image" src="https://github.com/MarkBind/markbind/assets/30100720/7693a24a-a495-4614-a0f5-976e60c39c50">

<img width="243" alt="image" src="https://github.com/MarkBind/markbind/assets/30100720/2d83bacf-06b2-4e57-bea5-fa872a47e58b">

Reviewer: Makes sense, wont need it to be in the icon t all.

> I strongly oppose converting the list behavior to a flex view. In my opinion, a list should function as a list unless specified otherwise.

> It would be more effective to specify an attribute like {flex} to achieve the desired presentation.

> Did you have something different in mind?

Was this layout (top and bottom) what you meant by `flex` as well? I cannot imagine how else you would position the icon and image (top and bottom vs left and right).

Author: > Was this layout (top and bottom) what you meant by flex as well? 

Refer to this: https://www.w3schools.com/css/css3_flexbox_responsive.asp

> I cannot imagine how else you would position the icon and image (top and bottom vs left and right).

The behavior of the list is oriented towards the left (icon) and right. So, simply use the icon attribute. If you prefer a top and bottom arrangement, remove the icon and focus on the content. This will nicely arrange items top and bottom on smaller screen sizes.

Reviewer: > Refer to this: https://www.w3schools.com/css/css3_flexbox_responsive.asp

Yes, but how would you use this to achieve something different from "top and bottom vs left and right"?

Author: Not sure what you mean?

Reviewer: > However, such an approach could disrupt the list presentation, which may not be ideal. It would be more effective to specify an attribute like {flex} to achieve the desired presentation.

What's the UI layout for the icon and list item content you were thinking of here?

Author: This represents a list:
Icon | Content
Icon | Content
Icon | Content

However, the following represents a card view layout similar to [Ant Design Cards](https://ant.design/components/card):
Icon
Content
Icon
Content
Icon
Content

Switching from a list to a card view layout may result in unexpected alignment issues. Imagine a scenario with a small icon and extensive text:

On a large screen, it might look like this:
Small icon: very long text repeated numerous times

But on a smaller screen, the alignment might change dramatically:
Small icon:
Very long text repeated numerous times

Clearly, the change in layout can cause significant alignment issues.

In addition, by maintaining the layout of a standard list, we can eventually achieve a UI layout like this:
<img width="408" alt="image" src="https://github.com/MarkBind/markbind/assets/30100720/abb71d6e-0807-441b-b664-a6789ad9b914">

Moreover, currently, we are crafting an icon list. However, in the future, you might need a navigation list, a grid list, and so forth. Hence, it's advisable to follow the standard list format unless specified otherwise by an attribute.

Reviewer: Sorry, I've only just gotten around to reading the `flex` part of your code.

There's likely some miscommunication here.

> However, such an approach could disrupt the list presentation, which may not be ideal. It would be more effective to specify an attribute like {flex} to achieve the desired presentation.

I understood from the conversation flow that you are suggesting `flex` as a solution to the user restrictiveness imposed by `max-width`. I don't see how this can be done apart from adopting a completely different UI layout. (with `flex` as the implementation)

I think we are talking about the same thing - just whether having `max-width` or not will be ideal in terms of just sticking to `flex` in your current implementation, inside a conventional "icon on left content on right" layout.

I'll resolve this but let me know if I got this wrong.

<variable name="code">

* item 1 {icon="/images/deer.jpg" size="xs"}
* item 1 {icon="/images/deer.jpg" size="s"}
* item 2 {icon="/images/deer.jpg" size="m"}
* item 3 {icon="/images/deer.jpg" size="l"}
* item 4 {icon="/images/deer.jpg" size="xl"}
</variable>
</include>

</div>

<box type="tip" seamless>

You can craft a media list using Markdown's heading and paragraph syntax.
</box>
<div id="main-example-markbind">

<include src="codeAndOutput.md" boilerplate >
<variable name="highlightStyle">markdown</variable>
<variable name="code">

* ### item 1
  
   This is the content for item 1. 
   Here we discuss the key aspects and details related to this specific item. 
   Information is comprehensive, well-organized, 
   and presented in a clear, concise manner.  
* ### item 2  

  For item 2, the content delves deeper into the topic, 
  providing a robust analysis and extensive insights. 
  The goal is to offer a thorough understanding of the subject matter.

* ### item 3 

  The content for item 3 takes a different angle, offering fresh perspectives and innovative ideas. 
  The objective is to stimulate thought and inspire creative solutions.

> {icon="/images/deer.jpg" size="xl" class="rounded"}

Author: I've modified the sizing to accept pixels (px) as I believe the attribute 'size' is more appropriate. This is because we are adjusting not only the width but also the height.

> {icon="/images/deer.jpg" size="xl" class="rounded"}

Reviewer: Should we be allowing specifying width and height separately?

16:9 images for example cannot be compressed into 1:1 nicely.

Naming nit: `icon-size/width/height/class` instead to make the attributes from this feature clear too. (e.g. not the list item class, but the icon's class)

> {icon="/images/deer.jpg" size="xl" class="rounded"}

Reviewer: related: https://github.com/MarkBind/markbind/pull/2316#discussion_r1252088583

> {icon="/images/deer.jpg" size="xl" class="rounded"}

Reviewer: > Naming nit: icon-size/width/height/class instead to make the attributes from this feature clear too. (e.g. not the list item class, but the icon's class)

@SPWwj if you missed this.

Would recommend `icon-class` at least since `class=` normally would apply to the list item. ([e.g.](https://github.com/MarkBind/markbind/pull/2316#discussion_r1264699066))

But not too sure if having a shorter form for `size/width/height` would be worth it over the consistency and ambiguity (`icon-size/width/height`), something we can get @damithc's user input on.

> {icon="/images/deer.jpg" size="xl" class="rounded"}

Reviewer 2: I'm fine to have a `icon-` prefix (or `i-`) for names to make it clear we are styling the icon.

> {icon="/images/deer.jpg" size="xl" class="rounded"}

Author: Yes, I see the conflict with the default attributes and have already updated to the `i-` naming convention.

<variable name="code">

* Item 1 {icon="/images/deer.jpg" width="30px" height="17px"}
* Item 2 {icon="/images/deer.jpg" width="60px" height="34px"}
* Item 3 {icon="/images/deer.jpg" width="90px" height="51px"}
* Item 4 {icon="/images/deer.jpg" width="120px" height="68px"}
* Item 5 {icon="/images/deer.jpg" width="150px" height="84px"}

</variable>
</include>

The `Width` and `height` is use to specific the dimension for the image icon, similar to size it support all css size units: [CSS Units](https://www.w3schools.com/cssref/css_units.php)
If the width and height are unspecified, the image's default dimensions will be used.

**The icon's appearance can be further customized by adding a `class` attribute.**

<div id="main-example-markbind">

<include src="codeAndOutput.md" boilerplate >
<variable name="highlightStyle">markdown</variable>
<variable name="code">
* Item 1 <br>
  Item 1 line 2{icon="/images/deer.jpg" width="60px" height="60px" height="17px" class="rounded"}
* Item 2
  Item 2 Continue
  * Item 2.1 {icon="fas-code-branch" class="text-success"}
  * Item 2.2 {class="text-danger"}
  * Item 2.3 
* Item 3
  * Item 3.1 
  * Item 3.2 {class="text-primary"}

>   * Item 3.3 {icon="fas-file-code" class="text-warning"}

Author: This is feasible. The syntax for such a list would be {text="Step 2", class="text-warning"}. Since we can't reuse the icon attribute, the text will collide with the emoji name.

>   * Item 3.3 {icon="fas-file-code" class="text-warning"}

Reviewer 2: Sound good. As mentioned, we can leave it as a future enhancement.

  * Item 3.4
* Item 4
  * Item 4.1 {icon="x" class="badge border border-primary rounded-pill mb-1 border-danger"}
  * Item 4.2 {icon="exclamation"}
</variable>

> </include>

Author: The screenshot looks odd to me. I can't replicate the appearance of your Item 4.2 in both Edge and Chrome browsers.<img width="240" alt="image" src="https://github.com/MarkBind/markbind/assets/30100720/8543d058-a367-43a1-be6b-05924a2ba929">

> On a related not, I wonder if alignment should be different for images (which are typically filled squares) compared to regular icons (which have transparent areas around them).

We can align the content vertically in the center and horizontally to the left. This layout appears most natural for a list.

> </include>

Reviewer 2: > We can align the content vertically in the center

I'm not sure about this. This works when the icon and the text is about the same sizes. For larger sizes, I think the content and the icon should align at the top.

> I can't replicate the appearance of your Item 4.2 in both Edge and Chrome browsers.

That's because emoji appearance depends on Browser/OS. So, don't use them where you want a consistent appearance. Use font based icons such as Fontawsome instead.

1. Once a class attribute is assigned to an icon, it will continue apply to icons at the same level, until it is overridden by a different class attribute (e.g., from Item 2.1 to Item 2.2, and so on to Item 3.2).

> 2. All standard Markdown rules will apply to the icon list (e.g., Item 1 and Item 2 to create a single item using multiple lines).

Reviewer 2: The more I think about it, the more I feel like this feature deserves its own name, such as `Fancy Lists`. Most other similar tools have something comparable, so we can name it as we wish.

> 2. All standard Markdown rules will apply to the icon list (e.g., Item 1 and Item 2 to create a single item using multiple lines).

Author: The icon list is an extension to the current unordered lists. Adding a new section for the icon list might not be visually appealing in the page navigation.

<img width="185" alt="image" src="https://github.com/MarkBind/markbind/assets/30100720/cc139b18-5463-48e8-b15e-2c040d4ab7c6">

> 2. All standard Markdown rules will apply to the icon list (e.g., Item 1 and Item 2 to create a single item using multiple lines).

Reviewer 2: It need not appear as a separate item. We can just replace `Customizing the list appearance`  with `Fancy Lists` (without making it a pageNav heading) and use that term from that point to refer to customized lists. 

More generally, either we choose a name for this feature and use it consistently throughout, or not use a name at all, referring to the feature as 'customizing' instead. Currently, we throw in the term `icon list` suddenly in the middle of the feature description.

- [x] Item 4
- ( ) Item 5
</div>

****Customizing the list appearance:****

**To customize list icons, add the configuration `{icon="icon-name"}` either after a specific list item or at the end of the entire list.**

<include src="codeAndOutput.md" boilerplate >
<variable name="highlightStyle">markdown</variable>
<variable name="code">

> * item 1 {icon="fas fa-times"}

Reviewer: > To customize list icons, add the configuration `{icon="icon-name"}` either after a specific list item or at the end of the entire list.

👀 Are we not aiming to maintain the existing icon syntax per this and your prior discussion? (without the `::`'s)

https://deploy-preview-2316--markbind-master.netlify.app/userguide/formattingcontents#icons

e.g.

- `fas fa-times` -> `fas-fa-times`
- `glyphicon glyphicon-education` -> `glyphicon-education`

p.s: I actually think removing the `::`'s here is a more intuitive option as its a delimiter used in markdown text, not as an attribute.

Reviewer: https://github.com/MarkBind/markbind/blob/master/packages/core/src/lib/markdown-it/plugins/markdown-it-icons.ts

Reviewer 2: @SPWwj as @ang-zeyu suggested, let's use the same icon syntax, minus the `:` delimiter.

Reviewer 2: > p.s: I actually think removing the `::`'s here is a more intuitive option as its a delimiter used in markdown text, not as an attribute.

@ang-zeyu Makes sense. That said, we already use `icon=":fas-home:"` syntax for boxes.

Author: 'fas fa-times' is a Bootstrap class, not an emoji. For users familiar with Bootstrap, it makes more sense to retain the original icon syntax 'fas fa-times' rather than converting it to a new syntax like 'fas-fa-times'.

The same goes for the emoji 'glyphicon glyphicon-education'.

Since these are all included in the official icon list, I believe there's no need for us to alter them.

Reviewer 2: @SPWwj we want to be as consistent as possible within MarkBind, before we try be consistent with Bootstrap. Icons are already an existing feature, and we should be able to say  "use the same syntax as [the icons feature]()" or failing that, "use the same syntax as [the icons feature]() but omit `:`"

Author: > @SPWwj we want to be as consistent as possible within MarkBind, before we try be consistent with Bootstrap. Icons are already an existing feature, and we should be able to say "use the same syntax as the icons feature" or failing that, "use the same syntax as the icons feature but omit `:`"

I will try to edit acordingly.

- [x] Item 4
- ( ) Item 5

> </div>

Reviewer 2: If committing image files, let's not use higher resolution than necessary, to reduce repo size.

****Customizing the list appearance:****

**To customize list icons, add the configuration `{icon="icon-name"}` either after a specific list item or at the end of the entire list.**

<include src="codeAndOutput.md" boilerplate >
<variable name="highlightStyle">markdown</variable>
<variable name="code">
* item 1 {icon="fas-fa-times"}
* item 2
  * item 2.1 {icon="fas-fa-check"}
  * item 2.2
* item 4 {icon="fas-fa-trophy"}

> * item 5 {icon="glyphicon-education"}

Reviewer 2: Add a multi-line item, to illustrate how to specify an icon for it?

* item 6 {icon="mif-perm-media"}
* item 7 {icon="zzz"}
* item 8 {icon="glyphicon-education"}
* item 9 {icon="octicon-git-pull-request"}
  </variable>
</include>

> 

Reviewer 2: I think at this point we can mention the rules applicable to the above example. We can't expect the reader to read until the end of this feature to find the rules. e.g.,

* Remember that Markdown (and hence, MarkBind) requires an empty line to precede any list.
* If an item consists of multiple lines, the icon specification should be at the end of the first line (see item x in the example above).
* ... other rules ...

**The icon's appearance can be further customized by adding a `class` attribute.**

<div id="main-example-markbind">

<include src="codeAndOutput.md" boilerplate >
<variable name="highlightStyle">markdown</variable>
<variable name="code">
* item 1 {icon="fas-fa-times" class="text-danger"}
* item 2 
  * item 2.1 {icon="fas-fa-check" class="text-primary"}
  * item 2.2
* item 3 {icon="fas-fa-home" class="text-warning"}
* item 4 {icon="fas-fa-trophy" class="text-warning"}
* item 5 {icon="glyphicon-education" class="text-warning"}
* item 6 {icon="mif-perm-media" class="text-warning"}
* item 7 {icon="zzz"}
* item 8 {icon="glyphicon-education" class="text-warning"}
* item 9 {icon="octicon-git-pull-request" class="text-warning"}
</variable>

> </include>

Reviewer 2: As the first example already shows how we can use many different types of icons, subsequent examples can limit the number of different icons used, so that the user can focus on the aspect being explained in that example. When explaining the `class` attribute, using a few font-awesome icons is best as they are easy to style using a `class`.

> </include>

Reviewer 2: This is a good place mention what happens if you set class but not the icon (and vice versa) to show that different attributes waterfall independently of each other (right?)

> </div>

Reviewer 2: Give an example of a lengthier class attribute e.g., `class="badge rounded-pill bg-danger text-white"`. Badges would look nice as list bullet points, but not sure if it will work in the current implementation.

On a related note, perhaps we can consider a `round`/`square` attributes to let users style the icon as a circle or a square (with a bg color)? No need to do in this version though.

**You can adjust the icon's size by using the `size` attribute.**

<include src="codeAndOutput.md" boilerplate >
<variable name="highlightStyle">markdown</variable>
<variable name="code">

* item 1 {icon="/images/deer.jpg" width="30px" height="17px"}
* item 2 {icon="/images/deer.jpg" width="60px" height="34px"}
* item 3 {icon="/images/deer.jpg" width="90px" height="51px"}
* item 4 {icon="/images/deer.jpg" width="120px" height="68px"}
* item 5 {icon="/images/deer.jpg" width="150px" height="84px"}

</variable>
</include>

> 

Reviewer 2: Here, you can mention any rules about specifying width/height. e.g., what if we omit one or both? Which units are allowed? Does it allow class attribute e.g., `class="rounded"` (to get rounded corners)

> 

Reviewer 2: The deploy preview does not show some of the expected outputs https://deploy-preview-2316--markbind-master.netlify.app/userguide/formattingcontents#code

![image](https://github.com/MarkBind/markbind/assets/1673303/6103ca1e-2b8d-4824-a92e-07123ad14490)
